packages/typespec-ts/src/modular/serialization/serializers.ts

```diff
@@ -0,0 +1,245 @@
+import { isDefined } from "@azure/core-util";
+import { UsageFlags } from "@typespec/compiler";
+import _ from "lodash";
+import { isSpecialHandledUnion } from "../buildSerializeUtils.js";
+import { Property, Type } from "../modularCodeModel.js";
+import {
+  Imports,
+  SerializerMap,
+  SerializerOutput,
+  getAllNamedAncestors,
+  getEncodingFormat,
+  getModularTypeId,
+  getParameterTypePropertyName,
+  getRLCTypeId,
+  getReturnTypePropertyName,
+  hasSerializeFunction,
+  isNormalUnion
+} from "./util.js";
+
+export function serializeType(
+  functionType: UsageFlags,
+  serializerMap: SerializerMap,
+  type: Type,
+  valueExpr: string
+): SerializerOutput {
+  if (hasSerializeFunction(type)) {
+    const modularTypeName = getModularTypeId(type) ?? "FIXMYNAME";
+    const serializerMetadata = serializerMap[modularTypeName];
+
+    const functionName =
+      functionType === UsageFlags.Input
+        ? serializerMetadata?.serializerFunctionName
+        : serializerMetadata?.deserializerFunctionName;
+    return {
+      expr: `${
+        functionName ?? functionType === UsageFlags.Input
+          ? "MISSING_SERIALIZER"
+          : "MISSING_DESERIALIZER"
+      }((${valueExpr}))`,
+      imports: functionName
+        ? [{ importType: "serializers", name: functionName }]
+        : undefined
+    };
+  }
+
+  switch (type.type) {
+    case "string":
+    case "constant":
+    case "float":
+    case "integer":
+    case "boolean":
+    case "any":
+      return { expr: valueExpr };
+    case "model": {
+      if (type.name) {
+        // has serialize function, should be unreachable
+        break;
+      }
+      return serializeModelInline(functionType, serializerMap, type, valueExpr);
+    }
+    case "datetime":
+      return { expr: serializeDatetime(functionType, type, valueExpr) };
+    case "list": {
+      return serializeList(functionType, serializerMap, type, valueExpr);
+    }
+    case "byte-array": {
+      return serializeByteArray(functionType, type, valueExpr);
+    }
+    case "combined":
+      if (isSpecialHandledUnion(type)) {
+        // has serialize function, should be unreachable
+        break;
+      }
+      return {
+        expr: isNormalUnion(type) ? valueExpr : `(${valueExpr}) as any`
+      };
+    case "enum":
+      // TODO: encodedName
+      return { expr: valueExpr };
+  }
+  return {
+    expr: "FIXME"
+  };
+}
+
+export function serializeByteArray(
+  functionType: UsageFlags,
+  type: Type,
+  valueExpr: string
+): SerializerOutput {
```

Comment:
I feel like this part might be problematic for streaming type, but since we haven't discussed the support of mfd in modular layer yet, I am not sure what should we do here.